### PR TITLE
More vehicle parts iterators:

### PR DESCRIPTION
**Pull request by BevapDin**
SUMMARY: None

- removed the `_at_relative` suffix of `parts_with_feature_at_relative` functions. The "at relative" part is unambiguous because of the additional parameter.

- changes some vehicle functions to not have default parameters. Those are difficult to read (some have `true` as default, some have `false`) and it's not clear whether `get_parts(flag)` is the same as `get_parts(flag, false, false)`.

Splits one function into several functions instead of having a boolean parameter (which is always a compile time constant!):

- `vehicle::get_parts(flag, true, false)` is now `vehicle::get_enabled_parts(flag)`

- `vehicle::get_parts(flag, false, true)` is now `vehicle::get_parts_including_broken(flag)`

- `vehicle::get_parts(flag, false, false)` is now `vehicle::get_parts(flag)`

- `vehicle::parts_with_feature(flag, false)` is now `vehicle::get_parts_including_broken(flag)`

- `vehicle::parts_with_feature(flag, true)` is now `vehicle::get_parts(flag)`

(There was no occurrence of `get_parts(true, true)` as that would yield parts that may be enabled *and* broken.)

Those new functions don't have any optional parameters and don't have any boolean parameters. They also return the new vehicle parts range class that can be iterated over and that expose a `vpart_reference`.

----

The next step can be seen at https://github.com/BevapDin/Cataclysm-DDA/commits/mrt

More functions will return the new vehicle parts range class and more properties of the `vehicle_part` can be accessed via `vpart_reference`.

I plan to make all properties of the vehicle parts accessible via the `vpart_reference`. The advantages is: it contains a `vehicle` references as well as the reference to the part itself. This allows to inform the `vehicle` object when properties a part have changed (e.g. the enabled status or the broken status), which can update its caches. It also makes the code shorter and easier to read.

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -8917,7 +8919,8 @@ bool game::plfire_check( const targeting_data &args )
         }
 
         if( gun->has_flag( "MOUNTED_GUN" ) ) {
-            const bool v_mountable = static_cast<bool>( m.veh_at( u.pos() ).part_with_feature( "MOUNTABLE" ) );
+            const bool v_mountable = static_cast<bool>( m.veh_at( u.pos() ).part_with_feature( "MOUNTABLE",
+                                     true ) );
```
Comment (alanbrady) [inline src/game.cpp]: Nitpick on this and the one below with the `true` var hanging off the newline, can I encourage:
```
const bool v_mountable =
    static_cast<bool>( m.veh_at( u.pos() ). part_with_feature( "MOUNTABLE", true ) )
```

File: src/map.cpp [hunk 2]
```diff
@@ -829,7 +829,8 @@ void map::board_vehicle( const tripoint &pos, player *p )
         return;
     }
 
-    const cata::optional<vpart_reference> vp = veh_at( pos ).part_with_feature( VPFLAG_BOARDABLE );
+    const cata::optional<vpart_reference> vp = veh_at( pos ).part_with_feature( VPFLAG_BOARDABLE,
+            true );
```
Comment (alanbrady) [inline src/map.cpp]: Ditto here

File: src/savegame_json.cpp [hunk 3]
```diff
@@ -2091,7 +2092,8 @@ void vehicle::deserialize( JsonIn &jsin )
     /** Legacy saved games did not store part enabled status within parts */
     auto set_legacy_state = [&]( const std::string & var, const std::string & flag ) {
         if( data.get_bool( var, false ) ) {
-            for( auto e : get_parts( flag ) ) {
+            for( const vpart_reference vp : get_parts( flag ) ) {
+                vehicle_part *const e = &vp.vehicle().parts[vp.part_index()];
```
Comment (alanbrady) [inline src/savegame_json.cpp]: Why pointer?

File: src/map_extras.cpp [hunk 4]
```diff
@@ -186,12 +189,13 @@ void mx_helicopter( map &m, const tripoint &abs_sub )
                 }
                 break;
             case 6: // Just pilots
-                for( auto p : wreckage->get_parts( VPFLAG_CONTROLS, false, true ) ) {
+                for( const vpart_reference vp : wreckage->get_parts_including_broken( VPFLAG_CONTROLS ) ) {
+                    const vehicle_part *const p = &vp.vehicle().parts[vp.part_index()];
```
Comment (alanbrady) [inline src/map_extras.cpp]: Any particular this is a pointer instead of a reference?

File: src/mission_companion.cpp [hunk 5]
```diff
@@ -3816,8 +3817,8 @@ std::string talk_function::camp_car_description( vehicle *car )
 {
     std::string entry = string_format( _( "Name:     %25s\n" ), car->name );
     entry += _( "----          Engines          ----\n" );
-    std::vector<vehicle_part *> part_engines = car->get_parts( "ENGINE" );
-    for( auto pt : part_engines ) {
+    for( const vpart_reference vp_ : car->get_parts( "ENGINE" ) ) {
+        const vehicle_part *const pt = &vp_.vehicle().parts[vp_.part_index()];
```
Comment (alanbrady) [inline src/mission_companion.cpp]: Why pointer?

File: src/map_extras.cpp (hunk 4 above)
Comment (BevapDin) [reply src/map_extras.cpp]: The current master already has this pointer variable. I just changed how it's calculated. This is going to be changed later.

Comment (alanbrady) [reply src/map_extras.cpp]: Yeah, it's just way more obvious now that's there's places where it's taking a pointer and then immediately converting it back.  If you're already planning on fixing it another PR then disregard. 
 Thanks.  Also thanks for cleaning up those `auto`s.

File: src/game.cpp (hunk 1 above)
Comment (BevapDin) [reply src/game.cpp]: Anyway, that line is going to change again later: I plan on removing the boolean parameter and change this to calla a function `part_with_feature_including_broken`, or similar. Maybe even just a `cargo_part` function without parameters (that returns broken parts).

**Comment by BevapDin** [comment]
> Did I merge this or mess up the branch?

Neither. You resolved the merge conflict by merging the current master into it.

**Comment by kevingranade** [comment]
To be clear, it's now appearing as containing no differences relative to master.

**Comment by ZhilkinSerg** [comment]
It seems to be merged:

default

**Comment by BevapDin** [comment]
Wonderful.

**Comment by ZhilkinSerg** [comment]
Maybe you have used Rebase and Merge button?

**Pull request closed without merging**